ssl/quic/quic_port.c

```diff
@@ -620,6 +620,90 @@ static int port_try_handle_stateless_reset(QUIC_PORT *port, const QUIC_URXE *e)
     return i > 0;
 }
 
+/**
+ * @brief Sends a QUIC Version Negotiation packet to the specified peer.
+ *
+ * This function constructs and sends a Version Negotiation packet using
+ * the connection IDs from the client's initial packet header. The
+ * Version Negotiation packet indicates support for QUIC version 1.
+ *
+ * @param port      Pointer to the QUIC_PORT structure representing the port
+ *                  context used for network communication.
+ * @param peer      Pointer to the BIO_ADDR structure specifying the address
+ *                  of the peer to which the Version Negotiation packet
+ *                  will be sent.
+ * @param client_hdr Pointer to the QUIC_PKT_HDR structure containing the
+ *                  client's packet header used to extract connection IDs.
+ *
+ * @note The function will raise an error if sending the message fails.
+ */
+static void port_send_version_negotiation(QUIC_PORT *port, BIO_ADDR *peer,
+                                          QUIC_PKT_HDR *client_hdr)
+{
+    BIO_MSG msg[1];
+    unsigned char buffer[512];
+    WPACKET wpkt;
+    QUIC_PKT_HDR hdr;
+    uint32_t supported_versions[1];
+    size_t written;
+
+    memset(&hdr, 0, sizeof(QUIC_PKT_HDR));
+    /*
+     * Reverse the source and dst conn ids
+     */
+    hdr.dst_conn_id = client_hdr->src_conn_id;
+    hdr.src_conn_id = client_hdr->dst_conn_id;
+
+    /*
+     * This is our list of supported protocol versions
+     * Currently only QUIC_VERSION_1
+     */
+    supported_versions[0] = QUIC_VERSION_1;
+
+    /*
+     * Fill out the header fields
+     * Note: Version negotiation packets, must, unlike
+     * other packet types have a version of 0
+     */
+    hdr.type = QUIC_PKT_TYPE_VERSION_NEG;
+    hdr.version = 0;
+    hdr.token = 0;
+    hdr.token_len = 0;
+    hdr.len = sizeof(supported_versions);
+    hdr.data = (unsigned char *)supported_versions;
+
+    msg[0].data = buffer;
+    msg[0].peer = peer;
+    msg[0].local = NULL;
+    msg[0].flags = 0;
+
+    if (!WPACKET_init_static_len(&wpkt, buffer, sizeof(buffer), 0))
+        return;
+
+    if (!ossl_quic_wire_encode_pkt_hdr(&wpkt, client_hdr->dst_conn_id.id_len,
+                                       &hdr, NULL))
+        return;
+
+    /*
+     * Add the array of supported versions to the end of the packet
+     */
+    if (!WPACKET_memcpy(&wpkt, supported_versions, sizeof(supported_versions)))
+        return;
+
+    if (!WPACKET_get_total_written(&wpkt, &msg[0].data_len))
+        return;
+
+    if (!WPACKET_finish(&wpkt))
+        return;
+
+    /*
+     * Send it back to the client attempting to connect
+     */
+    if (!BIO_sendmmsg(port->net_wbio, msg, sizeof(BIO_MSG), 1, 0, &written))
+        ERR_raise_data(ERR_LIB_SSL, SSL_R_QUIC_NETWORK_ERROR,
```

Comment:
I prefer to produce trace/log inetead of doing ERR_raise_data here.

Reply:
based on the stand up call today, I think we're going to implement this separately with qlog, tracked here:
https://github.com/openssl/project/issues/926